// react-native-reanimated/src/layoutReanimation/web/config.ts
'use strict';
import type { ReduceMotion, StyleProps } from 'react-native-reanimated/src/commonTypes';
import type { LayoutAnimationType } from 'react-native-reanimated/src/layoutReanimation/animationBuilder/commonTypes';
import {
  BounceIn,
  BounceInData,
  BounceOut,
  BounceOutData,
} from 'react-native-reanimated/src/layoutReanimation/web/animation/Bounce.web';
import { FadeIn, FadeInData, FadeOut, FadeOutData } from 'react-native-reanimated/src/layoutReanimation/web/animation/Fade.web';
import { FlipIn, FlipInData, FlipOut, FlipOutData } from 'react-native-reanimated/src/layoutReanimation/web/animation/Flip.web';
import {
  LightSpeedIn,
  LightSpeedInData,
  LightSpeedOut,
  LightSpeedOutData,
} from 'react-native-reanimated/src/layoutReanimation/web/animation/Lightspeed.web';
import { Pinwheel, PinwheelData } from 'react-native-reanimated/src/layoutReanimation/web/animation/Pinwheel.web';
import { RollIn, RollInData, RollOut, RollOutData } from 'react-native-reanimated/src/layoutReanimation/web/animation/Roll.web';
import {
  RotateIn,
  RotateInData,
  RotateOut,
  RotateOutData,
} from 'react-native-reanimated/src/layoutReanimation/web/animation/Rotate.web';
import {
  SlideIn,
  SlideInData,
  SlideOut,
  SlideOutData,
} from 'react-native-reanimated/src/layoutReanimation/web/animation/Slide.web';
import {
  StretchIn,
  StretchInData,
  StretchOut,
  StretchOutData,
} from 'react-native-reanimated/src/layoutReanimation/web/animation/Stretch.web';
import { ZoomIn, ZoomInData, ZoomOut, ZoomOutData } from 'react-native-reanimated/src/layoutReanimation/web/animation/Zoom.web';

import type { AnimationData, AnimationStyle } from 'react-native-reanimated/src/layoutReanimation/web/animationParser';

export type AnimationCallback = ((finished: boolean) => void) | null;

export type KeyframeDefinitions = Record<number, AnimationStyle>;

export type InitialValuesStyleProps = Omit<StyleProps, 'opacity'> & {
  opacity?: number;
};

export interface AnimationConfig {
  animationName: string;
  animationType: LayoutAnimationType;
  duration: number;
  delay: number;
  easing: string;
  callback: AnimationCallback;
  reversed: boolean;
}

interface EasingType {
  (): number;
  [EasingNameSymbol: symbol]: string;
}

export interface CustomConfig {
  easingV?: EasingType;
  easingXV?: EasingType;
  easingYV?: EasingType;
  durationV?: number;
  delayV?: number;
  randomizeDelay?: boolean;
  reduceMotionV?: ReduceMotion;
  callbackV?: AnimationCallback;
  reversed?: boolean;
  definitions?: KeyframeDefinitions;
  enteringV?: any;
  exitingV?: any;
  initialValues?: StyleProps;
}

export enum TransitionType {
  LINEAR,
  SEQUENCED,
  FADING,
  JUMPING,
  CURVED,
  ENTRY_EXIT,
}

export const AnimationsData: Record<string, AnimationData> = {
  ...FadeInData,
  ...FadeOutData,
  ...BounceInData,
  ...BounceOutData,
  ...FlipInData,
  ...FlipOutData,
  ...StretchInData,
  ...StretchOutData,
  ...ZoomInData,
  ...ZoomOutData,
  ...SlideInData,
  ...SlideOutData,
  ...LightSpeedInData,
  ...LightSpeedOutData,
  ...PinwheelData,
  ...RotateInData,
  ...RotateOutData,
  ...RollInData,
  ...RollOutData,
};

export const Animations = {
  ...FadeIn,
  ...FadeOut,
  ...BounceIn,
  ...BounceOut,
  ...FlipIn,
  ...FlipOut,
  ...StretchIn,
  ...StretchOut,
  ...ZoomIn,
  ...ZoomOut,
  ...SlideIn,
  ...SlideOut,
  ...LightSpeedIn,
  ...LightSpeedOut,
  ...Pinwheel,
  ...RotateIn,
  ...RotateOut,
  ...RollIn,
  ...RollOut,
};

export type AnimationNames = keyof typeof Animations;
export type LayoutTransitionsNames = keyof typeof AnimationsData;
